torch/csrc/api/src/nn/modules/transformer.cpp

```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
   return output;
 }
 
+
+// =======================================TransformerImpl================================
+TransformerImpl::TransformerImpl(TransformerOptions options_ ) : options(std::move(options_)){
+  reset();
+}
+
+void TransformerImpl::reset() {
+  // set up encoder
+  if (options.custom_encoder().is_empty()) {
+    LayerNorm norm(LayerNormOptions({options.d_model()}));
+    TransformerEncoder trans_encoder(TransformerEncoderOptions(
+      TransformerEncoderLayerOptions(options.d_model(), options.nhead())
+      .dim_feedforward(options.dim_feedforward())
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_encoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->encoder = AnyModule(trans_encoder);
+  }
+  else {
+    this->encoder = options.custom_encoder().clone();
+  }
+  this->register_module("encoder", this->encoder.ptr());
+
+  // set up decoder
+  if (options.custom_decoder().is_empty()) {
+
+    LayerNorm norm(LayerNormOptions({options.d_model()}));
+    TransformerDecoder trans_decoder(TransformerDecoderOptions(
+      TransformerDecoderLayerOptions(options.d_model(), options.nhead())
+      .dim_feedforward(options.dim_feedforward())
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_decoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->decoder = AnyModule(trans_decoder);
+  }
+  else {
+    this->decoder = options.custom_decoder().clone();
+  }
+  this->register_module("decoder", this->decoder.ptr());
+
+  reset_parameters();
+}
+
+
+void TransformerImpl::reset_parameters() {
+  auto parameters = this->parameters();
+  for (auto& param : parameters) {
+    if (param.dim() > 1) {
+      torch::nn::init::xavier_uniform_(param);
+    }
+  }
+}
+
+
+Tensor TransformerImpl::forward(
+  const Tensor& src,
+  const Tensor& tgt,
+  const Tensor& src_mask,
+  const Tensor& tgt_mask,
+  const Tensor& memory_mask,
+  const Tensor& src_key_padding_mask,
+  const Tensor& tgt_key_padding_mask,
+  const Tensor& memory_key_padding_mask) {
+
+  TORCH_CHECK(src.dim() == 3 && tgt.dim() == 3,
+    "src and tgt should have 3 dimensions, but got ", src.dim(), " and ", tgt.dim());
+
+  TORCH_CHECK(src.size(1) == tgt.size(1),
+    "src and tgt should have same size at dimension 1, but got ", src.size(1), " and ", tgt.size(1));
+
+  TORCH_CHECK(src.size(2) == options.d_model() && tgt.size(2) == options.d_model(),
+    "src and tgt should have a same size as d_model at dimension 2, but got ",
+    src.size(2), " and ", tgt.size(2), " while d_model is ", options.d_model());
```

Comment:
Nit: the python version mentions "feature" dimension in the error message which is more readable IMO: https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py#L121, it would be nice to include that as well as talk about dim 2